### PR TITLE
Wait_till option : normal

### PR DESCRIPTION
**Pull request by attilatabori**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
tabori@Attilas-MacBook-Pro terraform-provider-ibm % make testacc TEST=./ibm/service/kubernetes TESTARGS='-run=TestAccIBMContainerVpcClusterEnvvar'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./ibm/service/kubernetes -v -run=TestAccIBMContainerVpcClusterEnvvar -timeout 700m […] --- PASS: TestAccIBMContainerVpcClusterEnvvar (1560.43s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/kubernetes      1561.907s

data source with classic cluster:
tabori@Attilas-MacBook-Pro terraform-provider-ibm % make testacc TEST=./ibm/service/kubernetes TESTARGS='-run=TestAccIBMContainer_ClusterConfigDataSourceBasic'   
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./ibm/service/kubernetes -v -run=TestAccIBMContainer_ClusterConfigDataSourceBasic -timeout 700m […]
--- PASS: TestAccIBMContainer_ClusterConfigDataSourceBasic (1052.15s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/kubernetes      1053.532s

 vpc data source:  tabori@Attilas-MacBook-Pro terraform-provider-ibm % make testacc TEST=./ibm/service/kubernetes TESTARGS='-run=TestAccIBMContainer_ClusterConfigDataSourceVpcBasic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./ibm/service/kubernetes -v -run=TestAccIBMContainer_ClusterConfigDataSourceVpcBasic -timeout 700m
[…] 2022-11-08T10:56:30.313+0100 [DEBUG] sdk.helper_resource: Finished TestCase: test_name=TestAccIBMContainer_ClusterConfigDataSourceVpcBasic
--- PASS: TestAccIBMContainer_ClusterConfigDataSourceVpcBasic (1334.31s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/kubernetes      1335.692s

...
```

### CODE REVIEW

File: ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go [hunk 1]
```diff
@@ -246,8 +246,8 @@ func ResourceIBMContainerVpcCluster() *schema.Resource {
 				Optional:         true,
 				Default:          ingressReady,
 				DiffSuppressFunc: flex.ApplyOnce,
-				ValidateFunc:     validation.StringInSlice([]string{masterNodeReady, oneWorkerNodeReady, ingressReady}, true),
-				Description:      "wait_till can be configured for Master Ready, One worker Ready or Ingress Ready",
+				ValidateFunc:     validation.StringInSlice([]string{masterNodeReady, oneWorkerNodeReady, ingressReady, normal}, true),
+				Description:      "wait_till can be configured for Master Ready, One worker Ready or Ingress Ready or normal",
```
Comment (ocofaigh) [inline ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go:250]: wait_till can be configured for Master Ready, One Worker Ready, Ingress Ready or Normal

**Review by ocofaigh — COMMENTED** [review]
What does "normal" mean? And I guess docs need to be updated for this change too

**Comment by z0za** [comment]
> What does "normal" mean? And I guess docs need to be updated for this change too

"All worker nodes in a cluster are up and running. You can access the cluster and deploy apps to the cluster. This state is considered healthy and does not require an action from you."

https://cloud.ibm.com/docs/containers?topic=containers-cluster-states-reference#cluster-state-normal

yes, this needs to be added to the tf docs too.

**Comment by ocofaigh** [comment]
So what is going to be the difference between "Normal" and "IngressReady"?
According to the [docs](https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/resources/container_vpc_cluster):

> IngressReady: Terraform marks the creation of your cluster complete when the cluster master and all worker nodes are in a ready state, and the Ingress subdomain is fully set up.

Is it simply that the "Normal" will not wait until the Ingress subdomain is fully set up?

**Comment by z0za** [comment]
yes, that is my understanding

**Review by ocofaigh — COMMENTED** [review]
We should probably be consistent with naming here. The current values all use CamelCase (e.g. IngressReady) and looks like you are introducing a lower case value "normal". Should it be uppercase "Normal" ?

File: ibm/service/kubernetes/resource_ibm_container_cluster.go [hunk 2]
```diff
@@ -737,12 +737,22 @@ func resourceIBMContainerClusterCreate(d *schema.ResourceData, meta interface{})
 	if err != nil {
 		return err
 	}
-	if d.Get("wait_till").(string) == oneWorkerNodeReady {
+	waitForState := strings.ToLower(d.Get("wait_till").(string))
+
+	if waitForState == strings.ToLower(oneWorkerNodeReady) {
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_cluster.go:742]: maybe a switch-case would be better

File: ibm/service/kubernetes/resource_ibm_container_cluster.go [hunk 3]
```diff
@@ -1482,6 +1492,39 @@ func waitForClusterMasterAvailable(d *schema.ResourceData, meta interface{}) (in
...
+		return nil, err
+	}
+	csClient, err := meta.(conns.ClientSession).ContainerAPI()
+	if err != nil {
+		return nil, err
+	}
+	clusterID := d.Id()
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{deployRequested, deployInProgress},
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_cluster.go:1507]: maybe this should be parameter too. or... do we need a general function here?

File: ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go [hunk 4]
```diff
@@ -1216,6 +1222,40 @@ func waitForVpcClusterOneWorkerAvailable(d *schema.ResourceData, meta interface{
...
+	if err != nil {
+		return nil, err
+	}
+	csClient, err := meta.(conns.ClientSession).VpcContainerAPI()
+	if err != nil {
+		return nil, err
+	}
+	clusterID := d.Id()
+	createStateConf := &resource.StateChangeConf{
+		Pending: []string{deployRequested, deployInProgress},
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go:1236]: question as above

File: ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go [hunk 5]
```diff
@@ -1216,6 +1222,40 @@ func waitForVpcClusterOneWorkerAvailable(d *schema.ResourceData, meta interface{
...
+		Target:  []string{normal},
+		Refresh: func() (interface{}, string, error) {
+			clusterInfo, clusterInfoErr := csClient.Clusters().GetCluster(clusterID, targetEnv)
+
+			if err != nil || clusterInfoErr != nil {
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go:1241]: what is err here?

File: ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go [hunk 6]
```diff
@@ -1216,6 +1222,40 @@ func waitForVpcClusterOneWorkerAvailable(d *schema.ResourceData, meta interface{
...
+		return nil, err
+	}
+	csClient, err := meta.(conns.ClientSession).VpcContainerAPI()
+	if err != nil {
+		return nil, err
+	}
+	clusterID := d.Id()
+	createStateConf := &resource.StateChangeConf{
+		Pending: []string{deployRequested, deployInProgress},
+		Target:  []string{normal},
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go:1237]: if we go for a general solution, this should be waitForState

File: ibm/service/kubernetes/resource_ibm_container_vpc_cluster_test.go [hunk 7]
```diff
@@ -461,9 +461,10 @@ func testAccCheckIBMContainerVpcClusterEnvvar(name string) string {
 			subnet_id = "%[4]s"
 			name      = "us-south-1"
 		}
-		kms_instance_id = "%[5]s"
-		crk = "%[6]s"
-		kms_account_id = "%[7]s"
+		wait_till = "normal"
+		# kms_instance_id = "%[5]s"
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_vpc_cluster_test.go:465]: can we restore these?

File: ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go [hunk 8]
```diff
@@ -1216,6 +1222,40 @@ func waitForVpcClusterOneWorkerAvailable(d *schema.ResourceData, meta interface{
...
+			clusterInfo, clusterInfoErr := csClient.Clusters().GetCluster(clusterID, targetEnv)
+
+			if err != nil || clusterInfoErr != nil {
+				return clusterInfo, deployInProgress, clusterInfoErr
+			}
+
+			if clusterInfo.State == strings.ToLower(waitForState) {
+				return clusterInfo, waitForState, nil
+			}
+			return clusterInfo, deployInProgress, nil
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go:1248]: this means, we cannot fail fast. if the cluster ends up in a faulty state, we would wait till timeout...

File: website/docs/r/container_vpc_cluster.html.markdown [hunk 9]
```diff
@@ -198,7 +198,7 @@ Review the argument references that you can specify for your resource.
...
...ur  Terraform code can continue to run without waiting for the cluster to be fully created. Supported stages are: <ul><li><strong>`Normal`</strong>:  Terraform marks the creation of your cluster complete when the cluster is in a <code>normal</code> state. All worker nodes in a cluster are up and running. If you plan to do reading on the cluster from a datasource, use `Normal` </li><li><strong>`MasterNodeReady`</strong>:  Terraform marks the creation of your cluster complete when the cluster master is in a <code>ready</code> state.</li><li><strong>`OneWorkerNodeReady`</strong>:  Terraform marks the creation of your cluster complete when the master and at least one worker node are in a <code>ready</code> state.</li><li><strong>`IngressReady`</strong>:  Terraform marks the creation of your cluster complete when the cluster master and all worker nodes are in a <code>ready</code> state, and the Ingress subdomain is fully set up.</li></ul> If you do not specify this option, <code>`IngressReady`</code> is used by default. You can set this option only when the cluster is created. If this option is set during a cluster update or deletion, the parameter is ignored by the  Terraform provider.
```
Comment (z0za) [inline website/docs/r/container_vpc_cluster.html.markdown:201]: maybe we can shorten this description and point to the ibm cloud doc where one can read up about the states.
https://cloud.ibm.com/docs/containers?topic=containers-cluster-states-reference#cluster-state-normal

File: website/docs/r/container_cluster.html.markdown [hunk 10]
```diff
@@ -241,7 +241,7 @@ Review the argument references that you can specify for your resource.
...
+- `wait_till` - (Optional, String) The cluster creation happens in multi-stages. To avoid the longer wait times for resource execution.This argument in the resource will wait for the specified stage and complete the execution. The default stage value is `IngressReady`. The supported stages are  `MasterNodeReady` Resource waits till the master node is ready.  `OneWorkerNodeReady` Resource waits till one worker node is in to ready state. `Normal` Terraform marks the creation of your cluster complete when the cluster is in a [Normal](https://cloud.ibm.com/docs/containers?topic=containers-cluster-states-reference#cluster-state-normal) state. If you plan to do reading on the cluster from a datasource, use `Normal`. `IngressReady` Resource waits till the ingress-host and ingress-secret are available.
```
Comment (ocofaigh) [inline website/docs/r/container_cluster.html.markdown:244]: missing a space -> "execution.This"

**Pull request merged**